File: js/data_structure/linked_list.js
class LinkedList {
	constructor(values){
		if(arguments.length === 0){
			this.head=null;
			this.tail=null;
			this.length=0;
		}
		else{
			this.length = values.length;
			this.head = new LinkedList.Node(values[0]);
			let current = this.head;
			for(let i=1;i<this.length;i++) {
				current.next = new LinkedList.Node(values[i]);
				current = current.next;
			}
			this.tail = current;

		}
	}
	static Node = class {
		value = null;
		next = null;
		constructor(val=null,next=null){
			this.value=val;
			this.next=next;
		}
	}
	push(value){
		if(!this.head){
			this.head = new LinkedList.Node(value);
			this.tail = this.head;
		}
		else{
			let p = new LinkedList.Node(value);
			this.tail.next = p;
			this.tail = p;
		}
		this.length++;
	}
	get(idx){
		if (idx < 0 || idx> this.length){
			return undefined;
		}
		let current = this.head;
		let i = 0;
		while(i !== idx){
			current = current.next;
			i++;
		}
		return current;
	}
	pop(){
		if(!this.head){
			return undefined;
		}
		let popped = this.tail.value;
		if(this.length > 1) {
			this.tail = this.get(this.length - 2);
			this.tail.next = undefined;
		}
		else{
			this.head = undefined;
			this.tail = this.head;
		}
		this.length--;
		return popped;
	}

	/**
	 *
	 * @param idx Number
	 * @param value mixed
	 * @returns
	 */
	insert(idx,value){
		if(idx > this.length || idx < 0)
			return undefined;
		else if(idx === 0){
			this.unshift(value);
		}
		else if(idx === this.length){
			this.push(value);
		}
		else{
			let prev = this.get(idx - 1);
			prev.next = new LinkedList.Node(value, prev.next);
		}
		this.length++;
	}
	set(idx,value){
		if( idx > 0 && idx < this.length){
			let p = this.get(idx);
			p.value = value;
		}
	}
	shift(){
		if(this.head){
			let tmp = this.head;
			this.head = tmp.next;
			this.length--;
			if(this.length === 0){
				this.tail = undefined;
			}
			return tmp;
		}
	}
	unshift(value){
		if(this.head){
			this.head = new LinkedList.Node(value,this.head);
		}
		else{
			this.head = new LinkedList.Node(value);
			this.tail = this.head;
		}
		this.length++;
	}
	remove(idx){
		if(idx < 0 || idx > this.length){
			return undefined;
		}
		else if(idx === this.length - 1){
			this.pop();
		}
		else{
			let prev = this.get(idx-1);
			let remove = prev.next;
			prev.next = remove.next;
			this.length--;
		}
	}

	reverse(){
		let current = this.head;
		let prev = undefined;
		let next = undefined;
		while(current){
			next = current.next;
			current.next = prev;
			prev = current;
			current = next;
		}
		this.head = prev;
	}
	toString(){
		let current = this.head;
		let output = '['
		for(let i=0;i<this.length;i++){
			output += current.value.toString();
			if(i < this.length - 1){
				output += ', ';
			}
			current = current.next;
		}
		output += ']';
		return output;
	}
	display(){
		console.log(this.toString());
	}
}
function test(){
	const ll = new LinkedList([4,5]);
	ll.push(6);
	ll.insert(2,222);
	ll.push(7);
	ll.push(8);
	ll.push(9);
	ll.push(10);
	ll.reverse();
	ll.pop();
	ll.set(2,100);
	ll.insert(5,555);
	ll.insert(4,444);
	ll.shift();
	ll.unshift(6);
	ll.remove(5);
	const expected = '[6, 9, 100, 7, 444, 555, 222, 5]';
	if(ll.toString() === expected ){
		console.log("All's clear");
	}
	else{
		console.log(`We got ${ll.toString()} but we expected ${expected}.`)
	}
}
if (typeof module !== 'undefined' && require.main === module) {
	test();
}

File: lua/linked_list/linked_list.lua
--uncomment if you haven't set your working directory to the starting folder
--package.path = package.path .. ";../?.lua"
require 'util'
--[[
Methods


Note: The indexes are based on 1 as this is lua
]]

LinkedList = Class{}

function LinkedList:init()
	self.Node = Class{}
	function self.Node:init(value,ref)
		self.value = value
		self.next = ref
	end
	self._head = nil
	self._length = 0
	self._tail = nil
end

function LinkedList:get(idx)
	if idx < 1 or idx > self._length then
		return nil
	end

	local cur = self._head

	for i=1,idx-1 do
		print(cur.value)
		cur = cur.next
	end
	return cur
end

function LinkedList:push(value)
	if self._head == nil then
		self._head = self.Node(value)
		self._tail = self._head
	else
		local cur = self._head
		while(cur.next ~= nil) do
			cur = cur.next
		end
		cur.next = self.Node(value)
		self._tail = cur.next
	end
	self._length = self._length +  1
end

function LinkedList:pop(idx)
	if self._head == nil then
		return nil
	end
	local popped
	if idx then
		local prev = self:get(idx-1)
		popped = prev.next
		prev.next = prev.next.next
	else
		popped = self._tail.value
		if self._length > 1 then
			local prev = self:get(self._length-1)
			print_r(prev)
			self._tail = prev
			self._tail.next = nil
		else
			self._tail = nil
		end
	end
	self._length = self._length - 1
	return popped
end

function LinkedList:reverse()
	local cur = self._head
	self._head = self._tail
	self._tail = cur
	local prev = nil

	while(cur ~= nil) do
		local next = cur.next
		cur.next = prev
		prev = cur
		cur = next
	end
	self._head = prev
end

function LinkedList:insert(idx,value)
	if 1 > idx or idx > self._length then
		return nil
	end
	local prev = self:get(idx)
	prev.next = self.Node(value, prev.next)
	self._length = self._length + 1
end

function LinkedList:set(idx,value)
	if idx < 1 or idx > self._length then
		return nil
	end
	local Node = self:get(idx)
	Node.value = value
end

function LinkedList:len()
	return self._length
end

function LinkedList:shift()
	if self._head then
		local tmp = self._head
		self._head = tmp.next
		self._length = self._length - 1
		if self._length == 0 then
			self._tail = nil
		end
		return tmp
	else
		return nil
	end
end

function LinkedList:remove(idx)
	if 1 < idx or idx > self._length then
		return nil
	end
	if idx == self._length - 1 then
		self:pop()
	elseif idx == 0 then
		self:shift()
	else
		local prev = self:get(idx-1)
		local removed = prev.next
		prev.next = removed.next
		self._length = self._length - 1
	end
end

function LinkedList:unshift(value)
	if self._head then
		self._head = self.Node(value, self._head)
	else
		self._head = self.Node(value)
		self._tail = self._head
	end
end

function LinkedList:stringify()
	local os = '['
	local cur = self._head
	while cur.next ~= nil do
		os = os .. tostring(cur.value) .. ","
		cur = cur.next
	end
	return os .. tostring(cur.value) .. "]"
end

if not pcall(debug.getlocal,4,1) then
	ll = LinkedList()
	ll:push(1)
	ll:push(2)
	ll:push(3)
	ll:push(4)
	ll:push(5)
	ll:push(6)
	print(ll:stringify())
	ll:pop()
	ll:pop()
	print(ll:stringify())
	ll:shift()
	ll:unshift(11)
	print(ll:stringify())
	ll:pop(2)
	print(ll:stringify())
	-- nil test
	print(ll:get(0))
end